esp/esp/users/forms/user_profile.py

```diff
@@ -59,8 +59,10 @@ def __init__(self, *args, **kwargs):
         super(UserContactForm, self).__init__(*args, **kwargs)
         if not Tag.getBooleanTag('request_student_phonenum', default=True):
             del self.fields['phone_day']
-        if not Tag.getBooleanTag('text_messages_to_students'):
+        if not Tag.getBooleanTag('text_messages_to_students') or not self.user.isStudent():
             del self.fields['receive_txt_message']
+        if self.user.getLastProfile().contact_user and not self.user.getLastProfile().contact_user.receive_txt_message:
```

Comment:
Hmm, I'm a little confused what you're trying to do here.  Don't we initialize the form to the values from the user's profile, if they have one?  Is this just something to do with TypedChoiceField being weird, and if so can we just fix up the existing `.initial` value rather than re-feching it from the old profile?

Reply:
Yeah, it doesn't play very well with the boolean model. I'll look into fixing up the .initial value.

Reply:
So, after some digging, it looks like when the TypedChoiceField reads in the values from the user's profile, it reads it as either `True` or `None` (even when the value is actually `False`).
Therefore, this seems like the only way to initialize it correctly. I can probably clean it up a little bit by using `self.user.getLastProfile().text_reminder`?
Unless you have any other ideas?

Reply:
Or maybe I'm misunderstanding "fix up the existing `.initial` value"

Reply:
To further explain, there's 3 different cases we need to distinguish:

-None (there's no previous profile data)
-False (the user previously selected "No")
-True (the user previously selected "Yes")

For some reason (which I don't really understand), the form reads in the False case as `None`, which means we can't distinguish between the first two cases without any other information (like checking the last profile which is what I'm currently doing).

Reply:
Interesting.  I still don't think using `getLastProfile()` here is right; we should be using whatever data is passed into the form, which often will be from `getLastProfile()` but may be [`getLastForProgram()`](https://github.com/learning-unlimited/ESP-Website/blob/benkraft.new-site/esp/esp/web/views/myesp.py#L204) or some such.  Probably best is to use the initial data that has been passed into the form -- if `self.fields['receive_txt_message'].initial` doesn't have the right value, we should still be able to grab it from `kwargs.get('initial', {}).get('receive_txt_message')`.  Or we could have the caller (e.g. where I linked above) do the munging, before the form gets to it, although that's a bit uglier.  But I think leaving it as-is is going to be the source of some nasty edge case bugs when the profile in question isn't `getLastProfile()`.

Reply:
Ah, missed your two later comments.  By "fix up the existing `.initial` value", my hope is that even if the form doesn't display things right, we can still distinguish those three cases by looking at `self.fields['receive_txt_message'].initial` and fixing it.  But if not, i think the initial arg should also have the right data.

Reply:
Ah, I wasn't aware of kwargs.get('initial', {}), and I bet that will work. I'll look into that ASAP.

Reply:
So...that didn't actually work. It's still giving me `None` or `True` from the kwargs. Still investigating...

Reply:
The model is correct, because it's stored as `True`/`False` if the user has filled out the form and `None` if the user has never filled out a profile.
So somewhere the `False` is getting converted to `None` as it's passed to the form?

Reply:
Ahah, it appears that the receive_txt_message argument is only passed to kwargs when it is `True`. It isn't passed when it is `False`, which seems weird.
Hence why `kwargs.get('initial', {}).get('receive_txt_message')` returns `None`.

Reply:
After much digging, it looks like the [method](https://github.com/learning-unlimited/ESP-Website/blob/main/esp/esp/users/models/__init__.py#L1794) that supplies the initial values for the form skips any values that are `False`.
I assume it was intended to remove any missing values?

Reply:
Huh, interesting.  Yeah, I'd guess so.  Probably it should just be checking for `None` and `''`, but not for `False`, so you could explicitly check for those.  Just make sure you test a few different configurations of the form to make sure it doesn't break anything else.